File: compilers/java/src/main/java/runar/compiler/passes/Dce.java
package runar.compiler.passes;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import runar.compiler.ir.anf.AddDataOutput;
import runar.compiler.ir.anf.AddOutput;
import runar.compiler.ir.anf.AddRawOutput;
import runar.compiler.ir.anf.AnfBinding;
import runar.compiler.ir.anf.AnfMethod;
import runar.compiler.ir.anf.AnfProgram;
import runar.compiler.ir.anf.AnfValue;
import runar.compiler.ir.anf.ArrayLiteral;
import runar.compiler.ir.anf.Assert;
import runar.compiler.ir.anf.BinOp;
import runar.compiler.ir.anf.BytesConst;
import runar.compiler.ir.anf.Call;
import runar.compiler.ir.anf.CheckPreimage;
import runar.compiler.ir.anf.DeserializeState;
import runar.compiler.ir.anf.GetStateScript;
import runar.compiler.ir.anf.If;
import runar.compiler.ir.anf.LoadConst;
import runar.compiler.ir.anf.LoadParam;
import runar.compiler.ir.anf.LoadProp;
import runar.compiler.ir.anf.Loop;
import runar.compiler.ir.anf.MethodCall;
import runar.compiler.ir.anf.RawScript;
import runar.compiler.ir.anf.UnaryOp;
import runar.compiler.ir.UnknownAnfKindError;
import runar.compiler.ir.anf.UpdateProp;

/**
 * Dead Code Elimination pass for ANF IR.
 *
 * <p>Removes bindings whose results are never referenced by other bindings,
 * preserving bindings with observable side effects (assert, update_prop,
 * check_preimage, add_output, add_raw_output, add_data_output, call,
 * method_call, raw_script). Iterates to a fixed point so transitively
 * dead bindings are also removed.
 *
 * <p>"Results" is plural on purpose (N-140). A binding does not only define
 * its own {@code name}: an {@code if} that merges branch locals also defines
 * every name in {@code results}, and both an {@code if} and a {@code loop}
 * define the names their nested bindings bind. Liveness used to test
 * {@code used.contains(b.name())} alone, so an {@code if} named {@code t9}
 * carrying {@code results ["a","b"]} — a name nothing ever references,
 * because callers reference {@code a} and {@code b} — was deleted whenever its
 * arms happened to be pure, and the merged locals kept their pre-branch
 * values. See {@code conformance/dce/live-if.test.ts}.
 *
 * <p>This module is the canonical, standalone DCE pass for the Java
 * compiler. It mirrors the Zig reference implementation in
 * {@code compilers/zig/src/passes/dce.zig}. The earlier inline
 * implementation in {@code AnfOptimize.java} has been surgically
 * extracted here.
 *
 * <p>Behaviour: byte-for-byte identical, at the time of that extraction, to
 * the previous inline DCE in {@code AnfOptimize.java}. Verified by the
 * conformance suite (cross-tier hex parity) and the unknown-kind
 * exhaustiveness tests.
 *
 * <p>N-140 is the one deliberate behaviour change since: liveness considers
 * the names a binding DEFINES, not only its own {@code name}.
 */
public final class Dce {

    private Dce() {}

    // ---------------------------------------------------------------
    // Public API
    // ---------------------------------------------------------------

    /** Eliminate dead bindings across every method in the program. */
    public static AnfProgram run(AnfProgram program) {
        List<AnfMethod> newMethods = new ArrayList<>(program.methods().size());
        for (AnfMethod m : program.methods()) {
            List<AnfBinding> cleaned = eliminateDead(m.body());
            newMethods.add(new AnfMethod(m.name(), m.params(), cleaned, m.isPublic()));
        }
        return new AnfProgram(program.contractName(), program.properties(), newMethods);
    }

    /**
     * Remove bindings whose results are never referenced. Public for
     * delegation from {@link AnfOptimize} which still calls this during
     * the EC rewrite's post-cleanup phase.
     */
    public static List<AnfBinding> eliminateDead(List<AnfBinding> body) {
        List<AnfBinding> current = body;
        while (true) {
            List<Set<String>> ownRefs = new ArrayList<>(current.size());
            Map<String, Integer> refCount = new HashMap<>();
            for (AnfBinding b : current) {
                Set<String> own = new HashSet<>();
                collectRefs(b.value(), own);
                ownRefs.add(own);
                for (String name : own) refCount.merge(name, 1, Integer::sum);
            }

            List<AnfBinding> kept = new ArrayList<>(current.size());
            boolean removed = false;
            for (int i = 0; i < current.size(); i++) {
                AnfBinding b = current.get(i);
                if (isReferencedExternally(b, ownRefs.get(i), refCount) || hasSideEffect(b.value())) {
                    kept.add(b);
                } else {
                    removed = true;
                }
            }
            if (!removed) return kept;
            current = kept;
        }
    }

    /**
     * Every SSA name a binding brings into scope: its own {@code name}, plus —
     * for the two nesting kinds — an {@code if}'s declared {@code results} (the
     * merged branch locals / property slots both arms leave behind) and the
     * names bound inside {@code then}, {@code else} and a {@code loop} body,
     * recursively.
     *
     * <p>{@code iterVar} is deliberately absent: it is the loop's own induction
     * variable, referenced only from inside the body, so counting it as defined
     * would make every non-trivial loop unconditionally live.
     */
    public static void collectDefinedNames(AnfBinding binding, Set<String> out) {
        out.add(binding.name());
        AnfValue v = binding.value();
        if (v instanceof If ifv) {
            for (String r : orEmpty(ifv.results())) out.add(r);
            for (AnfBinding b : orEmpty(ifv.thenBranch())) collectDefinedNames(b, out);
            for (AnfBinding b : orEmpty(ifv.elseBranch())) collectDefinedNames(b, out);
        } else if (v instanceof Loop loop) {
            for (AnfBinding b : orEmpty(loop.body())) collectDefinedNames(b, out);
        }
    }

    /**
     * Is any name this binding defines referenced by some OTHER binding?
     *
     * <p>{@code refCount} maps a name to the number of DISTINCT bindings
     * referencing it; {@code ownRefs} is this binding's own contribution.
     * Subtracting it is what keeps the rule from degenerating into "never
     * delete an {@code if} or a {@code loop}": an arm's bindings almost always
     * reference each other, and counting those self-references would make
     * every nesting node immortal.
     *
     * <p>For a non-nesting binding this is exactly the old
     * {@code used.contains(b.name())}: ANF has no self-reference, so
     * {@code ownRefs} never holds the binding's own name.
     */
    private static boolean isReferencedExternally(
        AnfBinding binding, Set<String> ownRefs, Map<String, Integer> refCount) {
        Set<String> defined = new HashSet<>();
        collectDefinedNames(binding, defined);
        for (String name : defined) {
            int external = refCount.getOrDefault(name, 0) - (ownRefs.contains(name) ? 1 : 0);
            if (external > 0) return true;
        }
        return false;
    }

    // ---------------------------------------------------------------
    // Core algorithm
    // ---------------------------------------------------------------

    /**
     * Collect every binding name referenced (transitively) by a value.
     * Mirrors {@code _collect_refs} in
     * {@code compilers/python/runar_compiler/frontend/dce.py}.
     */
    public static void collectRefs(AnfValue v, Set<String> used) {
        if (v instanceof LoadParam || v instanceof LoadProp || v instanceof GetStateScript) {
            return;
        }
        if (v instanceof LoadConst lc) {
            // Track @ref: aliases as references so we don't DCE their target.
            if (lc.value() instanceof BytesConst bs) {
                String hex = bs.hex();
                if (hex != null && hex.startsWith("@ref:")) {
                    used.add(hex.substring(5));
                }
            }
            return;
        }
        if (v instanceof BinOp b) {
            used.add(b.left());
            used.add(b.right());
            return;
        }
        if (v instanceof UnaryOp u) {
            used.add(u.operand());
            return;
        }
        if (v instanceof Call c) {
            if (c.args() != null) used.addAll(c.args());
            return;
        }
        if (v instanceof MethodCall mc) {
            used.add(mc.object());
            if (mc.args() != null) used.addAll(mc.args());
            return;
        }
        if (v instanceof If ifv) {
            used.add(ifv.cond());
            for (AnfBinding tb : orEmpty(ifv.thenBranch())) collectRefs(tb.value(), used);
            for (AnfBinding eb : orEmpty(ifv.elseBranch())) collectRefs(eb.value(), used);
            return;
        }
        if (v instanceof Loop lp) {
            for (AnfBinding lb : orEmpty(lp.body())) collectRefs(lb.value(), used);
            return;
        }
        if (v instanceof Assert a) {
            used.add(a.value());
            return;
        }
        if (v instanceof UpdateProp up) {
            used.add(up.value());
            return;
        }
        if (v instanceof CheckPreimage cp) {
            used.add(cp.preimage());
            return;
        }
        if (v instanceof DeserializeState ds) {
            used.add(ds.preimage());
            return;
        }
        if (v instanceof AddOutput ao) {
            if (ao.satoshis() != null) used.add(ao.satoshis());
            if (ao.stateValues() != null) used.addAll(ao.stateValues());
            if (ao.preimage() != null) used.add(ao.preimage());
            return;
        }
        if (v instanceof AddRawOutput ar) {
            if (ar.satoshis() != null) used.add(ar.satoshis());
            if (ar.scriptBytes() != null) used.add(ar.scriptBytes());
            return;
        }
        if (v instanceof AddDataOutput ad) {
            if (ad.satoshis() != null) used.add(ad.satoshis());
            if (ad.scriptBytes() != null) used.add(ad.scriptBytes());
            return;
        }
        if (v instanceof ArrayLiteral al) {
            if (al.elements() != null) used.addAll(al.elements());
            return;
        }
        if (v instanceof RawScript) {
            // Opaque byte span — no SSA operand refs.
            return;
        }
        // Exhaustiveness guard. Silently returning here would let DCE drop a
        // live binding because its refs went uncollected.
        throw new UnknownAnfKindError(v.kind(), "anf-optimize.collectRefs");
    }

    /**
     * Side-effect predicate. Mirrors the Python {@code _has_side_effect} list:
     * assert, update_prop, check_preimage, deserialize_state, add_output,
     * add_raw_output, add_data_output, if, loop, call, method_call.
     */
    public static boolean hasSideEffect(AnfValue v) {
        if (v instanceof Assert
            || v instanceof UpdateProp
            || v instanceof CheckPreimage
            || v instanceof DeserializeState
            || v instanceof AddOutput
            || v instanceof AddRawOutput
            || v instanceof AddDataOutput
            || v instanceof Call
            || v instanceof MethodCall
            || v instanceof RawScript) { // opaque byte span — DCE must never eliminate it
            return true;
        }
        if (v instanceof If ifv) {
            for (AnfBinding tb : orEmpty(ifv.thenBranch())) {
                if (hasSideEffect(tb.value())) return true;
            }
            for (AnfBinding eb : orEmpty(ifv.elseBranch())) {
                if (hasSideEffect(eb.value())) return true;
            }
            return false;
        }
        if (v instanceof Loop lp) {
            for (AnfBinding lb : orEmpty(lp.body())) {
                if (hasSideEffect(lb.value())) return true;
            }
            return false;
        }
        // Issue #109 (@embedAlways): a load_prop injected to force a readonly
        // field into the deployed locking script carries `preserve = true`, so
        // DCE must keep it even though nothing references it. Ordinary
        // load_props (preserve = false) remain freely eliminable. Mirrors
        // compilers/zig/src/passes/dce.zig.
        if (v instanceof LoadProp lp) {
            return lp.preserve();
        }
        // Pure values — safe for DCE to drop when unreferenced.
        if (v instanceof LoadParam
            || v instanceof LoadConst
            || v instanceof GetStateScript
            || v instanceof BinOp
            || v instanceof UnaryOp
            || v instanceof ArrayLiteral) {
            return false;
        }
        // Exhaustiveness guard. A silent `return false;` would cause DCE to
        // eliminate a new side-effecting ANF kind, producing scripts that
        // omit observable behavior.
        throw new UnknownAnfKindError(v.kind(), "anf-optimize.hasSideEffect");
    }

    private static <T> List<T> orEmpty(List<T> list) {
        return list == null ? List.of() : list;
    }
}

File: packages/runar-compiler/src/optimizer/dce.ts
/**
 * Dead Code Elimination pass for ANF IR.
 *
 * Removes bindings whose results are never referenced by other bindings,
 * preserving bindings with observable side effects (assert, update_prop,
 * check_preimage, add_output, add_raw_output, add_data_output, call,
 * method_call, raw_script) and any `if` / `loop` whose nested bindings carry
 * one. Iterates to a fixed point so transitively dead bindings are also
 * removed.
 *
 * "Results" is plural on purpose (N-140). A binding does not only define its
 * own `name`: an `if` that merges branch locals also defines every name in
 * `results`, and both an `if` and a `loop` define the names their nested
 * bindings bind. Liveness used to test `refs.has(binding.name)` alone, so an
 * `if` named `t9` carrying `results: ['a','b']` — a name nothing ever
 * references, because callers reference `a` and `b` — was deleted whenever its
 * arms happened to be pure. The merged locals then kept their pre-branch
 * values and the contract compiled to a locking script that cannot be
 * satisfied. See `conformance/dce/live-if.test.ts` for the executable proof;
 * the defect was identical in all seven tiers, so cross-tier byte parity held
 * the whole time.
 *
 * This module is the canonical DEFINITION of DCE — one implementation, mirrored
 * by `compilers/zig/src/passes/dce.zig` and its five peers, and re-exported by
 * `optimizer/constant-fold.ts` (where it used to live inline) to preserve that
 * module's public API.
 *
 * It is NOT a standalone pipeline pass, and the header used to say it was
 * (R-194). The only caller is `optimizer/anf-ec.ts`, at the end of `optimizeEC`
 * and AFTER its `if (!anyChanged) return program;` early exit — so a program
 * the EC optimizer does not touch is never DCE'd. Go's `anf_optimize.go` does
 * the same thing, so there is no cross-tier divergence; the claim was simply
 * false.
 *
 * That gate is load-bearing, not an oversight. Measured by moving the call
 * ahead of the early exit so DCE runs on every program: the TypeScript tier
 * then FAILS TO COMPILE 11 of the 78 conformance fixtures —
 * all-readonly-cleanstack, bounded-loop, branch-merged-locals,
 * function-patterns, if-else, if-without-else, loop-if-merged-locals,
 * loop-shapes, merge-locals-prop-updates, merge-locals-shapes, multi-method.
 * Not byte movement: outright compilation failure. So this pass, run
 * unconditionally, removes bindings that stack lowering still needs, and
 * "make DCE standalone" is a defect to fix in DCE before it is a wiring change
 * (filed as N-140).
 *
 * Pinned by `r194-dce-invocation.test.ts` so this description and the call site
 * cannot drift apart again.
 *
 * Behaviour: byte-for-byte identical, at the time of that extraction, to the
 * previous in-place DCE inside `constant-fold.ts`. Verified by the conformance
 * suite (cross-tier hex parity) and the optimizer unit tests.
 *
 * N-140 is the one deliberate behaviour change since: liveness considers the
 * names a binding DEFINES, not only its own `name`.
 */

import type {
  ANFProgram,
  ANFMethod,
  ANFBinding,
  ANFValue,
} from '../ir/index.js';
import { PRESERVE } from '../ir/index.js';
import { UnknownANFKindError } from 'runar-ir-schema';

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Eliminate dead bindings across every method in the program.
 * Returns a new program; the input is not mutated.
 */
export function eliminateDeadBindings(program: ANFProgram): ANFProgram {
  return {
    ...program,
    methods: program.methods.map(eliminateDeadInMethod),
  };
}

function eliminateDeadInMethod(method: ANFMethod): ANFMethod {
  const live = filterLiveBindings(method.body);
  return { ...method, body: live };
}

// ---------------------------------------------------------------------------
// Core algorithm
// ---------------------------------------------------------------------------

export function collectRefsFromValue(value: ANFValue, refs: Set<string>): void {
  switch (value.kind) {
    case 'load_param':
    case 'load_prop':
    case 'get_state_script':
      break;
    case 'load_const':
      // Track @ref: aliases as references to prevent DCE
      if (typeof value.value === 'string' && value.value.startsWith('@ref:')) {
        refs.add(value.value.slice(5));
      }
      break;
    case 'bin_op':
      refs.add(value.left);
      refs.add(value.right);
      break;
    case 'unary_op':
      refs.add(value.operand);
      break;
    case 'call':
      for (const arg of value.args) refs.add(arg);
      break;
    case 'method_call':
      refs.add(value.object);
      for (const arg of value.args) refs.add(arg);
      break;
    case 'if':
      refs.add(value.cond);
      for (const b of value.then) collectRefsFromValue(b.value, refs);
      for (const b of value.else) collectRefsFromValue(b.value, refs);
      break;
    case 'loop':
      for (const b of value.body) collectRefsFromValue(b.value, refs);
      break;
    case 'assert':
      refs.add(value.value);
      break;
    case 'update_prop':
      refs.add(value.value);
      break;
    case 'check_preimage':
      refs.add(value.preimage);
      break;
    case 'deserialize_state':
      refs.add(value.preimage);
      break;
    case 'add_output':
      refs.add(value.satoshis);
      for (const sv of value.stateValues) refs.add(sv);
      refs.add(value.preimage);
      break;
    case 'add_raw_output':
      refs.add(value.satoshis);
      refs.add(value.scriptBytes);
      break;
    case 'add_data_output':
      refs.add(value.satoshis);
      refs.add(value.scriptBytes);
      break;
    case 'array_literal':
      for (const elem of value.elements) refs.add(elem);
      break;
    case 'raw_script':
      // Opaque: no SSA operand refs.
      break;
    default: {
      const unknown = value as { kind: string };
      throw new UnknownANFKindError(unknown.kind, 'constant-fold.collectRefsFromValue');
    }
  }
}

export function hasSideEffect(value: ANFValue): boolean {
  switch (value.kind) {
    case 'assert':
    case 'update_prop':
    case 'check_preimage':
    case 'deserialize_state':
    case 'add_output':
    case 'add_raw_output':
    case 'add_data_output':
    case 'call':        // calls may have side effects (e.g. assert)
    case 'method_call': // method calls may have side effects
    case 'raw_script':  // opaque byte span — DCE must never eliminate it
      return true;
    // `if` / `loop` are effectful iff some NESTED binding is. Nested bindings
    // live inside the parent node rather than flattened into the method body,
    // so retention is all-or-nothing: dropping an unreferenced `if` would take
    // every nested `assert` / `check_preimage` / `add_output` with it. Mirrors
    // the Go tier's `frontend.HasSideEffect` and the Rust tier's
    // `frontend::dce::has_side_effect`.
    case 'if':
      return (
        value.then.some((b) => hasSideEffect(b.value)) ||
        value.else.some((b) => hasSideEffect(b.value))
      );
    case 'loop':
      return value.body.some((b) => hasSideEffect(b.value));
    // Issue #109 (`@embedAlways`): a `load_prop` injected to force a readonly
    // field into the deployed locking script carries `[PRESERVE]: true`, so
    // DCE must keep it even though nothing references it. Ordinary load_props
    // leave the flag unset and remain freely eliminable. Mirrors
    // `compilers/zig/src/passes/dce.zig`.
    case 'load_prop':
      return value[PRESERVE] === true;
    // Pure ANF kinds — no side effect, safe to DCE if unreferenced.
    case 'load_param':
    case 'load_const':
    case 'get_state_script':
    case 'bin_op':
    case 'unary_op':
    case 'array_literal':
      return false;
    default: {
      const unknown = value as { kind: string };
      throw new UnknownANFKindError(unknown.kind, 'constant-fold.hasSideEffect');
    }
  }
}

/**
 * Every SSA name this binding brings into scope.
 *
 * Its own `name`, plus — for the two nesting kinds — the names it defines on
 * behalf of the enclosing method: an `if`'s declared `results` (the merged
 * branch locals / property slots both arms leave behind) and the names bound
 * inside `if.then`, `if.else` and `loop.body`, recursively.
 *
 * `loop.iterVar` is deliberately NOT here. It is the loop's own induction
 * variable, referenced only from inside the body, so counting it as defined
 * would make every non-trivial loop unconditionally live.
 */
export function collectDefinedNames(binding: ANFBinding, out: Set<string>): void {
  out.add(binding.name);
  collectDefinedNamesFromValue(binding.value, out);
}

function collectDefinedNamesFromValue(value: ANFValue, out: Set<string>): void {
  if (value.kind === 'if') {
    for (const r of value.results ?? []) out.add(r);
    for (const b of value.then) collectDefinedNames(b, out);
    for (const b of value.else) collectDefinedNames(b, out);
  } else if (value.kind === 'loop') {
    for (const b of value.body) collectDefinedNames(b, out);
  }
}

/**
 * Is `binding` referenced from OUTSIDE itself?
 *
 * `refCount` maps a name to the number of DISTINCT bindings that reference it;
 * `ownRefs` is the set this binding contributes. Subtracting its own
 * contribution is what keeps the rule from degenerating into "never delete an
 * `if` or a `loop`": an arm's bindings almost always reference each other, and
 * counting those self-references would make every nesting node immortal. Only
 * a reference from a sibling — the `a + b` after the branch — keeps it alive.
 *
 * For a non-nesting binding this is exactly the old `refs.has(binding.name)`:
 * ANF has no self-reference, so `ownRefs` never contains the binding's own
 * name and the subtraction is a no-op.
 */
function isReferencedExternally(
  binding: ANFBinding,
  ownRefs: Set<string>,
  refCount: Map<string, number>,
): boolean {
  const defined = new Set<string>();
  collectDefinedNames(binding, defined);
  for (const name of defined) {
    const external = (refCount.get(name) ?? 0) - (ownRefs.has(name) ? 1 : 0);
    if (external > 0) return true;
  }
  return false;
}

function filterLiveBindings(bindings: ANFBinding[]): ANFBinding[] {
  // Iterate to a fixed point so transitively dead bindings are removed too.
  let current = bindings;
  let changed = true;

  while (changed) {
    changed = false;
    const ownRefs = current.map((b) => {
      const s = new Set<string>();
      collectRefsFromValue(b.value, s);
      return s;
    });
    const refCount = new Map<string, number>();
    for (const s of ownRefs) {
      for (const name of s) refCount.set(name, (refCount.get(name) ?? 0) + 1);
    }
    const filtered: ANFBinding[] = [];

    for (let i = 0; i < current.length; i++) {
      const binding = current[i]!;
      if (
        isReferencedExternally(binding, ownRefs[i]!, refCount) ||
        hasSideEffect(binding.value)
      ) {
        filtered.push(binding);
      } else {
        changed = true;
      }
    }

    current = filtered;
  }

  return current;
}
